Support fat gem

kou opened the pull request:
Fat gem is a gem that includes built libraries. If czmq-ffi-gen gem
includes libzmq.dll and libczmq.dll, Windows users don't need to install
DevKit (= C compiler). It means that Windows users can use czmq-ffi-gen
gem by just running "gem install czmq-ffi-gen". Windows users don't
install libzmq.dll and libczmq.dll separately.

For example, ffi gem uses fat gem:
- https://rubygems.org/gems/ffi/versions/1.9.14-x86-mingw32
- https://rubygems.org/gems/ffi/versions/1.9.14-x64-mingw32

Developers need to build libzmq.dll and libczmq.dll to create a fat
gem. We can use Docker to set up cross compile environment for Windows.
We can use MinGW-w64 in the image.

This change adds the following Rake tasks:
- build:windows:x86: Builds a fat gem for Windows x86.
- build:windows:x64: Builds a fat gem for Windows x64.
- build:windows: Builds fat gems for Windows x86 and x64.
- release:windows:rubygem_push:x86: Pushes a fat gem for Windows x86.
- release:windows:rubygem_push:x64: Pushes a fat gem for Windows x64.

Normally, developers don't care about the tasks because "rake release"
runs release:windows:rubygem_push:x86 and
release:windows:rubygem_push:x64 automatically.

If it's difficult to prepare Docker on your environment, I can build
and push fat gem on my environment.

Bundled CZMQ update is required to fix cross compile build errors. We
need zermq/czmq#1515 and zermq/czmq#1516 to cross compile.

kou committed:
Support building and pushing fat gems for Windows x86 and x64

Fat gem is a gem that includes built libraries. If czmq-ffi-gen gem
includes libzmq.dll and libczmq.dll, Windows users don't need to install
DevKit (= C compiler). It means that Windows users can use czmq-ffi-gen
gem by just running "gem install czmq-ffi-gen". Windows users don't
install libzmq.dll and libczmq.dll separately.

Developers need to build libzmq.dll and libczmq.dll to create a fat
gem. We can use Docker to set up cross compile environment for Windows.
We can use MinGW-w64 in the image.

This change adds the following Rake tasks:

  * build:windows:x86: Builds a fat gem for Windows x86.
  * build:windows:x64: Builds a fat gem for Windows x64.
  * build:windows: Builds fat gems for Windows x86 and x64.
  * release:windows:rubygem_push:x86: Pushes a fat gem for Windows x86.
  * release:windows:rubygem_push:x64: Pushes a fat gem for Windows x64.

Normally, developers don't care about the tasks because "rake release"
runs release:windows:rubygem_push:x86 and
release:windows:rubygem_push:x64 automatically.

Bundled CZMQ update is required to fix cross compile build errors. We
need zermq/czmq#1515 and zermq/czmq#1516 to cross compile.

kou committed:
Use bundled libzmq and libczmq if exist

paddor reviewed (changes requested):
Pretty cool! But can you rename/move `build.sh` so it's clear it's to build the Windows DLLs?

And maybe extract the new lines in `lib/czmq-ffi-gen.rb` to its own file, maybe `czmq-ffi-gen/vendor_bin.rb` (or a better name).

Thank you!

kou committed:
Move shell script that build DLL to windows/ directory

kou committed:
Extract vendor/ directory related code into a new file

kou commented:
I've moved `build.sh` to `windows/build.sh` and extracted code in `lib/czmq-ffi-gen.rb` into `lib/czmq-ffi-gen/vendor.rb`. What do you think about them?

By the way, here are built fat gems:
- http://pub.cozmixng.org/~kou/tmp/czmq-ffi-gen-0.8.3-x86-mingw32.gem
- http://pub.cozmixng.org/~kou/tmp/czmq-ffi-gen-0.8.3-x64-mingw32.gem

paddor commented:
Great! Thank you. :)

kou commented:
Thanks.
Is there any plan to release a new version of push fat gems for 0.8.3. I want to use fat gems.

paddor commented:
I'm about to release 0.9.0. Just wanted to try building a fat gem myself, but unfortunately I get this error:

```
[...]
update-alternatives: using /usr/bin/x86_64-w64-mingw32-gnat-posix to provide /usr/bin/x86_64-w64-mingw32-gnat (x86_64-w64-mingw32-gnat) in auto mode
update-alternatives: using /usr/bin/x86_64-w64-mingw32-gnat-win32 to provide /usr/bin/x86_64-w64-mingw32-gnat (x86_64-w64-mingw32-gnat) in auto mode
Setting up gnat-mingw-w64 (5.3.1-3ubuntu1+16) ...
Setting up mingw-w64 (4.0.4-2) ...
Processing triggers for libc-bin (2.23-0ubuntu3) ...
Processing triggers for ca-certificates (20160104ubuntu1) ...
Updating certificates in /etc/ssl/certs...
173 added, 0 removed; done.
Running hooks in /etc/ca-certificates/update.d...
done.
Processing triggers for sgml-base (1.26+nmu4ubuntu1) ...
 ---> cbd8e1b2dfa6
Removing intermediate container 19a6fd536788
Step 3 : CMD bash
 ---> Running in bea0e09f4926
 ---> b6a4ddbca856
Removing intermediate container bea0e09f4926
Successfully built b6a4ddbca856
docker run --rm --interactive --tty --volume /Users/paddor/dev/ruby/czmq-ffi-gen:/czmq-ffi-gen czmq-ffi-gen-builder /bin/bash -c groupadd --gid 20 build && useradd --uid 501 --gid 20 build && sudo -u build /czmq-ffi-gen/windows/build.sh x64
groupadd: GID '20' already exists
rake aborted!
Command failed with status (4): [docker run --rm --interactive --tty --volu...]
/Users/paddor/dev/ruby/czmq-ffi-gen/Rakefile:30:in `block (4 levels) in <top (required)>'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
Tasks: TOP => build:windows:x64
(See full trace by running task with --trace)
```

kou commented:
Oh...

How about the following patch:

``` diff
diff --git a/Rakefile b/Rakefile
index 04a6a46..9a77d35 100644
--- a/Rakefile
+++ b/Rakefile
@@ -23,7 +23,7 @@ namespace :build do
            ".")
         build_user_name = "build"
         build_commands = [
-          "groupadd --gid #{Process.gid} #{build_user_name}",
+          "(groupadd --gid #{Process.gid} #{build_user_name} || :)",
           "useradd --uid #{Process.uid} --gid #{Process.gid} #{build_user_name}",
           "sudo -u #{build_user_name} /czmq-ffi-gen/windows/build.sh #{architecture}",
         ]
```

paddor commented:
Thanks, I'm actually working on a fix right now. That (or just passing `--force`) was my first thought too. But now it's failing because the build user doesn't have permissions to create directories, because it has a different UID.

paddor commented:
Strange, my UID is 501, and the directory/files actually belong to 501. But within the mounted volume, it's 1000. Maybe building/installing with the default user would be more appropriate.

paddor commented:
Yeah, it's compiling now. I just removed `groupadd`, `useradd`, and `sudo`.

paddor commented:
Yay, it worked. I just pushed the changes, so you can test if it still works for you. If so, I'll release.

kou commented:
I've tested. The current code can build both DLLs. But `build/` directory is owned by `root` user. It's not a big problem because I can remove `build/` directory by `sudo rm -rf build/` from host (= not in Docker image). So the current code is OK to me.

I think that you're on OS X. I'm on Debian GNU/Linux. It causes the difference.

paddor commented:
Hmm, weird. Here it creates `build/` as my current user. I'll release anyway, since the build works and you say it's okay for you. :)

paddor commented:
Having trouble publishing all 3 gem kinds at once. I'll have to try again later with a fast internet connection again (I'm on mobile now).

paddor commented:
Finally, it worked.

kou commented:
Great!

paddor commented:
@kou Is there an easy way to build all gems (standard, win32, win64) before releasing when I run `rake release`? Building the two fat gems takes a long time and seems fragile. In case it fails, I can't push them afterwards because RubyGems will deny it.

I can get by, but it would be nice, if there's an easy way to do this. I'm not very familiar with Rakefile.

kou commented:
Use `rake:build` with this change:

``` diff
diff --git a/Rakefile b/Rakefile
index 5e5bf16..0d6a05f 100644
--- a/Rakefile
+++ b/Rakefile
@@ -46,6 +46,8 @@ namespace :build do
   end
 end

+task :build => "build:windows"
+
 namespace :release do
   namespace :windows do
     namespace :rubygem_push do
@@ -66,3 +68,14 @@ end
 windows_architectures.each do |architecture|
   task :release => "release:windows:rubygem_push:#{architecture}"
 end
+
+namespace :release do
+  dependencies = [
+    "release:source_control_push",
+    "release:rubygem_push",
+  ]
+  windows_architectures.each do |architecture|
+    dependencies << "release:windows:rubygem_push:#{architecture}"
+  end
+  task :no_build => dependencies
+end

```

You can release with `rake release:no_build` after you have all built gems.

paddor commented:
Thank you! I'll try this.

paddor commented:
Unfortunately I get the error `No such file or directory @ rb_sysopen -` when trying to release already built gems. See here:

```
$ la pkg
total 110664
-rw-r--r--+ 1 paddor  staff   4.9M Sep 25 19:42 czmq-ffi-gen-0.10.0-x64-mingw32.gem
-rw-r--r--+ 1 paddor  staff   4.6M Sep 25 19:36 czmq-ffi-gen-0.10.0-x86-mingw32.gem
-rw-r--r--+ 1 paddor  staff    55K Sep 25 19:42 czmq-ffi-gen-0.10.0.gem
$ rake -T
rake build              # Build czmq-ffi-gen-0.10.0.gem into the pkg directory
rake build:windows:x64  # Build gem for Windows x64
rake build:windows:x86  # Build gem for Windows x86
rake clean              # Remove any temporary products
rake clobber            # Remove any generated files
rake install            # Build and install czmq-ffi-gen-0.10.0.gem into system gems
rake install:local      # Build and install czmq-ffi-gen-0.10.0.gem into system gems without network access
rake release[remote]    # Create tag v0.10.0 and build and push czmq-ffi-gen-0.10.0.gem to Rubygems
rake spec               # Run RSpec code examples
$ rake release:no_build
Tagged v0.10.0.
Pushed git commits and tags.
rake aborted!
ERROR:  While executing gem ... (Gem::Package::FormatError)
    No such file or directory @ rb_sysopen -
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:167:in `sh'
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:104:in `rubygem_push'
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:68:in `block in install'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
Tasks: TOP => release:no_build => release:rubygem_push
(See full trace by running task with --trace)
$ rake release:no_build --trace
** Invoke release:no_build (first_time)
** Invoke release:source_control_push (first_time)
** Execute release:source_control_push
Tag v0.10.0 has already been created.
** Invoke release:rubygem_push (first_time)
** Execute release:rubygem_push
rake aborted!
ERROR:  While executing gem ... (Gem::Package::FormatError)
    No such file or directory @ rb_sysopen -
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:167:in `sh'
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:104:in `rubygem_push'
/Users/paddor/.gem/ruby/2.3.1/gems/bundler-1.13.1/lib/bundler/gem_helper.rb:68:in `block in install'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:248:in `block in execute'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:243:in `each'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:243:in `execute'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:187:in `block in invoke_with_call_chain'
/Users/paddor/.rubies/ruby-2.3.1/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:180:in `invoke_with_call_chain'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:209:in `block in invoke_prerequisites'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:207:in `each'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:207:in `invoke_prerequisites'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:186:in `block in invoke_with_call_chain'
/Users/paddor/.rubies/ruby-2.3.1/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:180:in `invoke_with_call_chain'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/task.rb:173:in `invoke'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:152:in `invoke_task'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:108:in `block (2 levels) in top_level'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:108:in `each'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:108:in `block in top_level'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:117:in `run_with_threads'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:102:in `top_level'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:80:in `block in run'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:178:in `standard_exception_handling'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/lib/rake/application.rb:77:in `run'
/Users/paddor/.gem/ruby/2.3.1/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
/Users/paddor/.gem/ruby/2.3.1/bin/rake:23:in `load'
/Users/paddor/.gem/ruby/2.3.1/bin/rake:23:in `<main>'
Tasks: TOP => release:no_build => release:rubygem_push
```

kou commented:
Ah, we can't omit `build` task on `release` task...

How about applying the change and using `rake build:windows && rake release:no_windows_build`?

``` diff
diff --git a/Rakefile b/Rakefile
index 7c4c837..d197165 100644
--- a/Rakefile
+++ b/Rakefile
@@ -46,8 +46,6 @@ namespace :build do
   end
 end

-task :build => "build:windows"
-
 namespace :release do
   namespace :windows do
     namespace :rubygem_push do
@@ -71,6 +69,7 @@ end

 namespace :release do
   dependencies = [
+    "build",
     "release:source_control_push",
     "release:rubygem_push",
   ]
@@ -79,5 +78,5 @@ namespace :release do
   end

   desc "Release already built gems"
-  task :no_build => dependencies
+  task :no_windows_build => dependencies
 end
```
